control/grpc.py

```diff
@@ -268,6 +271,41 @@ def get_namespace_infos_for_anagrpid(self, nqn: str, anagrpid: int) -> Iterator[
                 if ns_info.anagrpid == anagrpid:
                     yield ns_info
 
+    def get_all_namespaces_by_ana_group_id(self, anagrpid):
+        ns_list = []
+        # Loop through all nqn values in the namespace list
+        for nqn in self.namespace_list:
+            for nsid in self.namespace_list[nqn]:
+                ns = self.namespace_list[nqn][nsid]
+                if ns.empty():
+                    continue
+                if ns.anagrpid == anagrpid:
+                    ns_list.append((nsid, nqn))#list of tupples
+        return ns_list
+
+    def get_ana_group_id_by_nsid_subsys(self, nqn, nsid):
+        if nqn not in self.namespace_list:
+            return 0
```

Comment:
Yes. After you tested that nqn is in the list, you need to test that nsid is in the namespace_list[nqn]. Otherwise you'll get an exception if nsid is not there.